math/minuit2/src/FumiliBuilder.cxx

```diff
@@ -263,19 +289,187 @@ FunctionMinimum FumiliBuilder::Minimum(const MnFcn &fcn, const GradientCalculato
       }
 
+      if (doTrustRegion) {
+         // do simple trust region using some control delta and eta
+         // compute norm of Newton step
+         double norm = 0;
+         if (scaleTR) {
+            print.Debug("scaling Trust region with diagonal matrix D ",D);
+            stepScaled = D * step;
+            norm = sqrt(inner_product(stepScaled, stepScaled) );
+         }
+         else
+            norm = sqrt(inner_product(step,step));
+         // some conditions
+
+         // double tr_radius = norm;
+         // if (norm < 0.1 * delta) tr_radius = 0.1*delta;
+         // else if (norm > delta) tr_radius = delta;
+
+         // update new point using reduced step
+         //step = (tr_radius/norm) * step;
+
+         // if the proposed point (newton step) is inside the trust region radius accept it
+         if (norm <= delta) {
+            p = MinimumParameters(s0.Vec() + step, fcn(s0.Vec() + step));
+            print.Debug("Accept full Newton step - it inside TR ",delta);
+         } else {
+            //step = - (delta/norm) * step;
+
+            // if Newton step  is outside try to use the Cauchy point ?
+            double normGrad2 = 0;
+            double gHg = 0;
+            if (scaleTR) {
+               auto gScaled = Dinv * s0.Gradient().Grad();
+               normGrad2 = inner_product(gScaled, gScaled);
+               // compute D-1 H D-1 = H(i,j) * D(i,i) * D(j,j)
+               MnAlgebraicSymMatrix Hscaled(n);
+               for (unsigned int i = 0; i < n; i++) {
+                  for (unsigned int j = 0; j <=i; j++) {
+                     Hscaled(i,j) = H(i,j) * Dinv(i,i) * Dinv(j,j);
+                  }
+               }
+               gHg = similarity(gScaled, Hscaled);
+            } else {
+               normGrad2 = inner_product(s0.Gradient().Grad(),s0.Gradient().Grad());
+               gHg = similarity(s0.Gradient().Grad(), s0.Error().Hessian());
+            }
+            double normGrad = sqrt(normGrad2);
+            // need to compute gTHg :
+
+
+            print.Debug("computed gdel gHg and normGN",gdel,gHg,norm*norm, normGrad2);
+            if (gHg <= 0.) {
+               // Cauchy point is at the trust region boundary)
+               step = - (delta/ normGrad) * s0.Gradient().Grad();
+               if (scaleTR) step = Dinv2 * step;
+               print.Debug("Use as new point the Cauchy  point - along gradient with norm=delta ", delta);
+            } else {
+               // Cauchy point can be inside the trust region
+               double tau = std::min( (normGrad2*normGrad)/(gHg*delta), 1.0);
+               if (tau == 1.0) {
+                  // Cauchy is at the boundary
+                  step = - (delta/ normGrad) * s0.Gradient().Grad();
+                  if (scaleTR)
+                    step = Dinv2 * step;
+
+                  print.Debug("tau=1. - Use as new point the Cauchy  point - along gradient with norm=delta ", delta);
+               } else {
```

Comment:
Are some of the code paths not redundant here? You have two code branches for where the Cauchy point is at the boundary: like 360 and line 368 (with `tau == 1.0`). Maybe the second branch is redundant? That would also avoid code repetition.